web3/_utils/encoding.py

```diff
@@ -309,3 +313,19 @@ def encode_single_packed(_type, value):
         return codecs.encode(value, 'utf8')
     elif abi_type.base == "bytes":
         return value
+
+
+class Web3JsonEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, AttributeDict):
+            return {k: v for k, v in obj.items()}
```

Comment:
This seems like it should be a recursive operation, re-encoding each of the `k` and `v` values using `json.JSONEncoder.default(self, v)`

Reply:
Looks like `json.JSONEncoder` already handles recursion.  `default()` just needs to provide one to one conversion.  I added a test with a nested `AttributeDict` as verification.